src/lib.rs

```diff
@@ -727,15 +727,219 @@ impl CertificateParams {
 		let alg = SignatureAlgorithm::from_oid(&alg_oid.collect::<Vec<_>>())?;
 
 		let dn = DistinguishedName::from_name(&x509.tbs_certificate.subject)?;
+		let is_ca = Self::convert_x509_is_ca(&x509)?;
+		let validity = x509.validity();
+		let serial = Self::convert_x509_serial(&x509);
+		let subject_alt_names = Self::convert_x509_subject_alternative_name(&x509)?;
+		let key_usages = Self::convert_x509_key_usages(&x509)?;
+		let extended_key_usages = Self::convert_x509_extended_key_usages(&x509)?;
+		let name_constraints = Self::convert_x509_name_constraints(&x509)?;
+
 		Ok(
 			CertificateParams {
 				alg,
+				is_ca,
+				subject_alt_names,
+				key_usages,
+				extended_key_usages,
+				name_constraints,
 				distinguished_name : dn,
 				key_pair : Some(key_pair),
+				serial_number: Some(serial),
+				not_before : validity.not_before.to_datetime(),
+				not_after : validity.not_after.to_datetime(),
 				.. Default::default()
 			}
 		)
 	}
+	#[cfg(feature = "x509-parser")]
+	fn convert_x509_is_ca(x509 :&x509_parser::certificate::X509Certificate<'_>) -> Result<IsCa, RcgenError> {
+		use x509_parser::extensions::BasicConstraints as B;
+
+		let basic_constraints = x509.basic_constraints()
+			.or(Err(RcgenError::CouldNotParseCertificate))?.map(|ext| ext.value);
+
+		let is_ca = match basic_constraints {
+			Some(B { ca: true, path_len_constraint: Some(n) }) => IsCa::Ca(BasicConstraints::Constrained(*n as u8)),
+			Some(B { ca: true, path_len_constraint: None }) => IsCa::Ca(BasicConstraints::Unconstrained),
+			Some(B { ca: false, .. }) => IsCa::ExplicitNoCa,
+			None => IsCa::NoCa,
+		};
+
+		Ok(is_ca)
+	}
+	#[cfg(feature = "x509-parser")]
+	fn convert_x509_serial(x509 :&x509_parser::certificate::X509Certificate<'_>) -> u64 {
+		const SIZE: usize = std::mem::size_of::<u64>();
+
+		let raw = x509.raw_serial();
+		let len = raw.len();
+
+		if len >= SIZE {
+			return u64::from_be_bytes(raw[(len - SIZE)..].try_into().unwrap())
```

Comment:
same here, I think an error (or ignoring the serial if it doesnt fit) is better than truncation. In other places we ignore stuff unknown to us, and that is fine, but there is a difference between "i don't know" and "i know what this is but can't fit the data so I truncate it".

Reply:
I've changed it to return an error in f29f839.

Reply:
This is now failing tests sometimes because the serial number is too large. According to RFC 5280, a serial number of up to 20 bytes must be handled, but that is currently impossible in rcgen due to the use of `u64` for this field.

Do you think it's ok to change the field type to [BigUint](https://docs.rs/num-bigint/0.4.3/num_bigint/struct.BigUint.html) (or maybe just keep the raw bytes, as `Vec<u8>`, possibly in a new field)?. Otherwise I don't see an option other than not parsing large serial numbers, but that's kinda unsatisfying.

Reply:
This is a tough question, I would like to keep this library simple but I also understand that we have to support 20 bytes or more. Maybe this could be deferred for later?

Reply:
On the one hand I understand not wanting to introduce a dependency, but on the other I guess it would be easier to break backward compatibility now rather than later.

What do you think about just adding a `raw_serial_number: Option<Vec<u8>>` field to `CertificateParams` for now?

Reply:
Do you have a specific use case for the serial number stuff? Otherwise I don't think we need to address this right now.

Reply:
I would like to use an UUID that identifies the certificate as its serial number for my use case. It's more of a "would be nice to have" thing though, not strictly necessary, as I can use some other field for that.

Reply:
Hmm this makes sense. Having a separate field feels like a hack to me. I think it should be the main serial number field, but we should use a new type wrapper for the big number so that we don't depend on that crate's API. Ideally something that contains the `Vec<u8>` directly without needing a crate, then offering conversion methods to/from integers. But this is a large design space. That's why I said I wanted to have a separate PR for it, as the other changes here are fine.

Reply:
I've removed serial number parsing in 5a40247. Hopefully I'll be able to work on a separate PR for that soon.